go/vt/discovery/tablet_health.go

```diff
@@ -0,0 +1,368 @@
+	conn := th.getConnection()
+	if conn == nil {
+		// This signals the caller to retry
+		return nil
+	}
+	err := conn.StreamHealth(ctx, callback)
+	if err != nil {
+		// Depending on the specific error the caller can take action
+		th.closeConnection(ctx, err)
+	}
+	return err
+}
+
+func (th *tabletHealthCheck) getConnection() queryservice.QueryService {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+	if th.Conn == nil {
+		conn, err := tabletconn.GetDialer()(th.Tablet, grpcclient.FailFast(true))
+		if err != nil {
+			th.LastError = err
+			return nil
+		}
+		th.Conn = conn
+		th.LastError = nil
+	}
+	return th.Conn
+}
+
+func (th *tabletHealthCheck) closeConnection(ctx context.Context, err error) {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+	log.Warningf("tablet %v healthcheck stream error: %v", th.Tablet.Alias, err)
+	th.setServingState(false, err.Error())
+	th.LastError = err
+	_ = th.Conn.Close(ctx)
+	th.Conn = nil
+}
+
+// processResponse reads one health check response, and updates health
+func (th *tabletHealthCheck) processResponse(hc *HealthCheckImpl, shr *query.StreamHealthResponse) error {
+	select {
+	case <-th.ctx.Done():
+		return th.ctx.Err()
+	default:
+	}
+
+	// Check for invalid data, better than panicking.
+	if shr.Target == nil || shr.RealtimeStats == nil {
+		return fmt.Errorf("health stats is not valid: %v", shr)
+	}
+
+	// an app-level error from tablet, force serving state.
+	var healthErr error
+	serving := shr.Serving
+	if shr.RealtimeStats.HealthError != "" {
+		healthErr = fmt.Errorf("vttablet error: %v", shr.RealtimeStats.HealthError)
+		serving = false
+	}
+
+	if shr.TabletAlias != nil && !proto.Equal(shr.TabletAlias, th.Tablet.Alias) {
+		// TabletAlias change means that the host:port has been taken over by another tablet
+		// We could cancel / exit the healthcheck for this tablet right away
+		// However, we defer it until the next topo refresh informs us of the change because that is
+		// the only way to discover the new host/port
+		return vterrors.New(vtrpc.Code_FAILED_PRECONDITION, fmt.Sprintf("health stats mismatch, tablet %+v alias does not match response alias %v", th.Tablet, shr.TabletAlias))
+	}
+
+	th.mu.Lock()
+	currentTarget := th.Target
+	// check whether this is a trivial update so as to update healthy map
+	trivialNonMasterUpdate := th.LastError == nil && th.Serving && shr.RealtimeStats.HealthError == "" && shr.Serving &&
+		currentTarget.TabletType != topodata.TabletType_MASTER && currentTarget.TabletType == shr.Target.TabletType
+	isMasterUpdate := shr.Target.TabletType == topodata.TabletType_MASTER
+	// Track how often a tablet gets promoted to master. It is used for
+	// comparing against the variables in go/vtgate/buffer/variables.go.
+	isMasterChange := currentTarget.TabletType != topodata.TabletType_MASTER && shr.Target.TabletType == topodata.TabletType_MASTER
+	th.mu.Unlock()
+
+	// hc.healthByAlias is authoritative, it should be updated
+	hc.mu.Lock()
```

Comment:
I think from here on to end should be moved to a function `HealthCheckImpl`.

Reply:
Done